### PR TITLE
[ROCm] No-fence in normalization kernel

### PR DESCRIPTION
**Pull request by anatoliylitv**
Removing need for fences in normalization kernel by converting the stores into atomics+return. This is crucial for perf in architectures with split caches (e.g. MI300), where fences are inherently costly. This change speedups `batch_norm_stats ` function for tensors in `channels_last` format.

### Performance result on MI300:
batchnorm_latency_comparison

**Particular Example:**
Before: 
Avg time for shape (20, 896, 59, 91): **1102.39 us**

After:
Avg time for shape (20, 896, 59, 91): **122.94 us**

Reproducer:
```

import torch

shapes = [(20, 896, 59, 91)]

eps = 1e-5

for shape in shapes:
    x = torch.randn(shape, device='cuda', dtype=torch.bfloat16)
    x = x.to(memory_format=torch.channels_last)
    for _ in range(20):
        _ = torch.batch_norm_stats(x, eps)
    torch.cuda.synchronize()

    start_evt = torch.cuda.Event(enable_timing=True)
    end_evt = torch.cuda.Event(enable_timing=True)
    start_evt.record()
    for _ in range(100):
        _ = torch.batch_norm_stats(x, eps)
    end_evt.record()
    torch.cuda.synchronize()
    print(f"Avg time for shape {shape}: {start_evt.elapsed_time(end_evt) / 100 * 1e3:.2f} us")

```

Related fix which is released: https://github.com/pytorch/pytorch/pull/161180

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

File: aten/src/ATen/native/cuda/KernelUtils.cuh [hunk 1]
```diff
@@ -252,13 +254,16 @@ __device__ inline void cmtdStore(void* address, T value) {
       else if constexpr (num_char_per_val*sizeof(char) == sizeof(value))
         for (int i=0; i<num_char_per_val; i++)
           __hip_atomic_store(reinterpret_cast<char *>(address)+i, _pnr.c[i], __ATOMIC_RELAXED, __HIP_MEMORY_SCOPE_AGENT);
-      __atomic_signal_fence(__ATOMIC_SEQ_CST);
+      if(wait_for_commit)
```
Comment (amd-hhashemi) [inline aten/src/ATen/native/cuda/KernelUtils.cuh]: I think it's better to make this wait_for_commit a templated arg. That way compiler will completely remove the waits when not needed. Having it as a runtime argument runs risk of compiler inserting a wait(0) here regardless (because at compile-time it needs to know the total number of possible inflight stores). I'm just asking because I've seen compiler do that in past.

**Review by amd-hhashemi — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cuda/KernelUtils.cuh (hunk 1 above)
Comment (jerrymannil) [inline aten/src/ATen/native/cuda/KernelUtils.cuh]: `if constexpr (wait_for_commit)`

**Review by jerrymannil — APPROVED** [review]

**Review by jeffdaily — APPROVED** [review]

**Review by amd-hhashemi — APPROVED** [review]
